ml-agents/mlagents/trainers/components/reward_signals/reward_signal.py

```diff
@@ -0,0 +1,68 @@
+import logging
+from mlagents.trainers.trainer import UnityTrainerException
+from mlagents.trainers.policy import Policy
+from collections import namedtuple
+import numpy as np
+import abc
+
+import tensorflow as tf
+
+logger = logging.getLogger("mlagents.trainers")
+
+RewardSignalResult = namedtuple(
+    "RewardSignalResult", ["scaled_reward", "unscaled_reward"]
+)
+
+
+class RewardSignal(abc.ABC):
+    def __init__(self, policy: Policy, strength, gamma):
+        """
+        Initializes a reward signal. At minimum, you must pass in the policy it is being applied to, 
+        the reward strength, and the gamma (discount factor.)
+        :param policy: The Policy object (e.g. PPOPolicy) that this Reward Signal will apply to. 
+        :param strength: The strength of the reward. The reward's raw value will be multiplied by this value. 
+        :param gamma: The time discounting factor used for this reward. 
+        :return: A RewardSignal object. 
+        """
+        class_name = self.__class__.__name__
+        short_name = class_name.replace("RewardSignal", "")
+        self.stat_name = f"Policy/{short_name} Reward"
+        self.value_name = f"Policy/{short_name} Value Estimate"
+        self.gamma = gamma
+        self.policy = policy
+        self.strength = strength
+
+    def evaluate(self, current_info, next_info):
+        """
+        Evaluates the reward for the agents present in current_info given the next_info
+        :param current_info: The current BrainInfo. 
+        :param next_info: The BrainInfo from the next timestep.
+        :return: a tuple of (scaled intrinsic reward, unscaled intrinsic reward) provided by the generator
+        """
+        return (
+            self.strength * np.zeros(len(current_info.agents)),
+            np.zeros(len(current_info.agents)),
+        )
+
+    def update(self, update_buffer, n_sequences):
+        """
+        If the reward signal has an internal model (e.g. GAIL or Curiosity), update that model.
+        :param update_buffer: An AgentBuffer that contains the live data from which to update.
+        :param n_sequences: The number of sequences in the training buffer.
+        :return: A dict of {"Stat Name": stat} to be added to Tensorboard
+        """
+        return {}
+
+    @classmethod
+    def check_config(cls, config_dict, param_keys=None):
```

Comment:
Should this be static ? Also, not only used in reward signals I think.

Reply:
It's not static so I can get the `cls` name. 

But yeah, there is similar logic in the Trainer (though slightly different), though it's weird to call a Trainer function in RewardSignals. I wonder if it's time to make a `utils.py` file with all of these common functions? Also some common logic with minibatching and sequence computing might go in this file.